test/functional/mempool_basic_reorg.py

```diff
@@ -0,0 +1,160 @@
+        'id': { 'BlockReward': '0x{}'.format(block_id) },
+        'index': index,
+    }
+
+def tx_input(tx_id, index = 0):
+    return {
+        'id': { 'Transaction': '0x{}'.format(tx_id) },
+        'index': index,
+    }
+
+def make_tx(inputs, output_amounts, time_lock = 0, flags = 0):
+    outputs = [ {'Transfer': [ { 'Coin': amt }, { 'AnyoneCanSpend': None } ]}
+               for amt in output_amounts ]
+    witness = { 'NoSignature': None }
+    tx = {
+        'version': 1,
+        'flags': flags,
+        'inputs': inputs,
+        'outputs': outputs,
+        'time_lock': time_lock,
+    }
+    signed_tx = {
+        'transaction': tx,
+        'signatures': [witness for _ in outputs],
+    }
+    tx_id = hash_object(base_tx_obj, tx)
+    encoded_tx = signed_tx_obj.encode(signed_tx).to_hex()[2:]
+    return (encoded_tx, tx_id)
+
+def make_empty_block(parent_id, nonce):
+    empty_merkle_root = "0x" + hash_object(vec_output_obj, [])
+    pow_data = {
+        'bits': 0x207fffff,
+        'nonce': nonce,
+    }
+    header = {
+        'version': 1,
+        'prev_block_id': "0x{}".format(parent_id),
+        'tx_merkle_root': empty_merkle_root,
+        'witness_merkle_root': empty_merkle_root,
+        'timestamp': int(time.time()),
+        'consensus_data': { 'PoW': pow_data },
+    }
+    signed_header = {
+        'header': header,
+        'signature': { 'None': None },
+    }
+    block = {
+        'header': signed_header,
+        'reward': [],
+        'transactions': [],
+    }
+    block_id = hash_object(block_header_obj, header)
+    encoded_block = block_obj.encode(block).to_hex()[2:]
+    return (encoded_block, block_id)
+
+def mine_empty_block(parent_id):
+    for nonce in range(1000):
+        (block, block_id) = make_empty_block(parent_id, nonce)
+        if block_id[-2] in "01234567":
+            return (block, block_id)
+    assert False, "Cannot mine block"
+
+class MempoolTxSubmissionTest(BitcoinTestFramework):
+
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [[]]
+
+    def setup_network(self):
+        self.setup_nodes()
+        self.sync_all(self.nodes[0:1])
+
+    def run_test(self):
+        node = self.nodes[0]
+
+        # Get chain tip
+        genesis_id = node.chainstate_best_block_id()
+        self.log.debug('Initial tip: {}'.format(genesis_id))
+
+        # Prepare three transactions, each spending the previous one in sequence
+        (tx1, tx1_id) = make_tx([ reward_input(genesis_id) ], [ 1_000_000 ] )
+        self.log.debug("Encoded tx1 {}: {}".format(tx1_id, tx1))
+        (tx2, tx2_id) = make_tx([ tx_input(tx1_id) ], [ 900_000 ] )
+        self.log.debug("Encoded tx2 {}: {}".format(tx2_id, tx2))
+        (tx3, tx3_id) = make_tx([ tx_input(tx2_id) ], [ 800_000 ] )
+        self.log.debug("Encoded tx3 {}: {}".format(tx3_id, tx3))
+
+        # Submit the first transaction
+        node.p2p_submit_transaction(tx1)
```

Comment:
The difference between submitting a transaction through p2p and submitting the transaction directly in the mempool, is that the latter doesn't broadcast. This seems to be superfluous.

Reply:
The RPC for that was not available so I added it and changed the mempool functional tests to use it.